### DISCUSSION TITLE
How do I pass a variable from extension.js to package.json?

### DISCUSSION THREAD

**Discussion by tbazant** [microsoft/vscode-discussions#621] [Extension Development QnA] [answered]
In my extension, i get the value i need in the `activate` function like this:
```
function activate(context) {
  vscode.window.showInformationMessage('Path to my extension is: ' + context.extensionPath);
}
```
How do i assign it to  variable so that i'm able to make use of it in the `package.json` MANIFEST file?

**Comment by Muhammadahmed23** [comment]
In general, you cannot directly pass a variable from extension.js to package.json since they serve different purposes. The package.json file is primarily used for configuring and describing your project, while extension.js typically contains code logic for your extension or module.

However, if you need to make use of a variable value defined in extension.js within your project configuration, such as in scripts or build processes specified in package.json, you have a few options:

Environment Variables: One approach is to set an environment variable in your project's build environment or development setup. You can define the variable value in extension.js and then access it in package.json using process.env.VARIABLE_NAME.

Configuration Files: Instead of passing a variable directly, you can use a configuration file (e.g., JSON, YAML) to store the variable value. You would read the value from extension.js and write it to the configuration file. Then, in package.json, you can read the value from the configuration file as part of your build or script commands.

Build Tools or Task Runners: If you are using a build tool or task runner like Gulp, Grunt, or Webpack, you can define custom tasks or plugins that read the variable from extension.js and use it during the build process. This allows you to integrate the variable value into your build or deployment pipeline without directly accessing package.json.

That's the specific implementation depends on your project's requirements and the build tools or frameworks you're using. It's important to understand the capabilities and conventions of the tools you're working with to effectively pass variables between different parts of your project.

**Reply by tbazant** [reply]
Thanks @Muhammadahmed23 , let's choose the environment variable approach :-)
In the `launch.json`, i've defined `suseValeConfigPath`:
```
"configurations": [
{
	"name": "Run Extension",
	"type": "extensionHost",
	"request": "launch",
	"args": [
		"--extensionDevelopmentPath=${workspaceFolder}"
	],
	"outFiles": [
		"${workspaceFolder}/out/**/*.js"
	],
	"preLaunchTask": "${defaultBuildTask}",
	"env": {
		"suseValeConfigPath": ".vale.ini"
	}
},
```
then in the `extension.js`, i'm doing"
```
function activate(context) {
    let suseValeConfigPath = context.extensionPath;
    vscode.window.showInformationMessage(suseValeConfigPath);
}
```
and in `package.json`:
```
"contributes": {
    "configurationDefaults": {
      "vale.valeCLI.config": "process.env.suseValeConfigPath/.vale.ini"
    }
  },
```

which was a nice try but does not work :-) Where exactly is the glitch please?

**Reply by r3m0t** [reply]
This seems AI generated.

**Comment by alefragnani** [comment]
Hi @tbazant ,

Could you provide more details _why_ do you need to update the `package.json` file from within your extension? What's the purpose of it? 

I mean, `package.json` is the _description/metadata_ of your extension, and should describe your extension upfront, instead of being updated every time your extension is activated. BTW, VS Code uses that info to know your extension and load it. If you update the value after it is activated, I suppose it won't be re-loaded, and the updated settings/values will be ignored.

With that in mind, I think what you need is _update the setting value_ in User Preferences, which you should use `workspace.getConfiguration("vale").update("valeCLI.config", someValue);` instead. 

Another thing that I couldn't understand correctly is if `vale.valeCLI.config` setting is from your extension. Since you are using `configurationDefaults`, I suppose it is not. Is that right?

Hope this helps

**Reply by tbazant** [reply, marked as answer]
Hi @Muhammadahmed23 

> 
> Could you provide more details _why_ do you need to update the `package.json` file from within your extension? What's the purpose of it?

i just need to override the `vale.valeCLI.config` option from another extension to point to a file at the root directory of my extension, that's all :-)


> 
> I mean, `package.json` is the _description/metadata_ of your extension, and should describe your extension upfront, instead of being updated every time your extension is activated. BTW, VS Code uses that info to know your extension and load it. If you update the value after it is activated, I suppose it won't be re-loaded, and the updated settings/values will be ignored.

understood

> 
> With that in mind, I think what you need is _update the setting value_ in User Preferences, which you should use `workspace.getConfiguration("vale").update("valeCLI.config", someValue);` instead.

exactly, will try it, thanks a lot for this!


> 
> Another thing that I couldn't understand correctly is if `vale.valeCLI.config` setting is from your extension. Since you are using `configurationDefaults`, I suppose it is not. Is that right?
 right, it is not

> 
> Hope this helps

it surely will

**Reply by tbazant** [reply]
```
const vscode = require("vscode");
function activate(context) {
    let suseValeConfigPath = context.extensionPath;
    let configuration = vscode.workspace.getConfiguration("vale");
    configuration.update("valeCLI.config", suseValeConfigPath + "/.vale.ini");
}
```

made the trick, thanks a lot
